bson/value_reader.go

```diff
@@ -85,54 +34,94 @@ type valueReader struct {
 	offset int64
 	d      []byte
 
+	readerErr error
+	r         io.Reader
+
 	stack []vrState
 	frame int64
 }
 
 // NewValueReader returns a ValueReader using b for the underlying BSON
-// representation. Parameter b must be a BSON Document.
-func NewValueReader(b []byte) ValueReader {
-	// TODO(skriptble): There's a lack of symmetry between the reader and writer, since the reader takes a []byte while the
-	// TODO writer takes an io.Writer. We should have two versions of each, one that takes a []byte and one that takes an
-	// TODO io.Reader or io.Writer. The []byte version will need to return a thing that can return the finished []byte since
-	// TODO it might be reallocated when appended to.
-	return newValueReader(b)
+// representation.
+func NewValueReader(r io.Reader) ValueReader {
+	return newValueReader(r)
 }
 
 // NewBSONValueReader returns a ValueReader that starts in the Value mode instead of in top
 // level document mode. This enables the creation of a ValueReader for a single BSON value.
-func NewBSONValueReader(t Type, val []byte) ValueReader {
-	stack := make([]vrState, 1, 5)
-	stack[0] = vrState{
-		mode:  mValue,
-		vType: t,
-	}
-	return &valueReader{
-		d:     val,
-		stack: stack,
+func NewBSONValueReader(t Type, r io.Reader) ValueReader {
+	vr := newValueReader(r)
+	if len(vr.stack) == 0 {
+		vr.stack = make([]vrState, 1, 5)
 	}
+	vr.stack[0].mode = mValue
+	vr.stack[0].vType = t
+	return vr
 }
 
-func newValueReader(b []byte) *valueReader {
+func newValueReader(r io.Reader) *valueReader {
 	stack := make([]vrState, 1, 5)
 	stack[0] = vrState{
 		mode: mTopLevel,
 	}
 	return &valueReader{
-		d:     b,
+		r:     r,
 		stack: stack,
 	}
 }
 
-func (vr *valueReader) reset(b []byte) {
-	if vr.stack == nil {
-		vr.stack = make([]vrState, 1, 5)
+func newValueReaderFromSlice(buf []byte) *valueReader {
+	vr := newValueReader(nil)
+	vr.d = buf
+	return vr
+}
+
+func (vr *valueReader) prepload(length int32) (int32, error) {
+	const chunckSize = 512
```

Comment:
Optional: There are a some simplifications that would make this method mostly unnecessary:
* The functionality of `prepload` has a lot of overlap with the `bufio.Reader` type. Consider refactoring `valueReader` to use methods provided by a `bufio.Reader` instead.
* The current code won't actually reduce memory usage by reading the document incrementally because we're still appending every byte read to the slice `vr.d`. A simpler approach would be to use `ReadDocument` to read the entire next BSON document from the input stream anytime we reach the end of `vr.d`, allowing the rest of the code to mostly stay the same.
For example:
```go
func (vr *valueReader) prepload() error {
	if offset < len(vr.d)
		return nil
	}

	doc, err := ReadDocument(vr.r)
	if err != nil {
		return err
	}
	vr.d = doc
	return nil
}
```

Reply:
`vr.d` kept in this PR is primarily for the compatibility of `newValueReaderFromSlice`, which is only used by three deprecated methods in `unmarshal.go`. I'm converting this PR to a draft and will update the implementation after merging PR #1698 where all deprecated code is removed.

Reply:
I will refine the implementation in a following PR.